File: libspi/hypertext.c
/*
 * AT-SPI - Assistive Technology Service Provider Interface
 * (Gnome Accessibility Project; http://developer.gnome.org/projects/gap)
 *
 * Copyright 2008 Novell, Inc.
 * Copyright 2001, 2002 Sun Microsystems Inc.,
 * Copyright 2001, 2002 Ximian, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "accessible.h"

static AtkHypertext *
get_hypertext (DBusMessage * message)
{
  AtkObject *obj = spi_dbus_get_object (dbus_message_get_path (message));
  if (!obj)
    return NULL;
  return ATK_HYPERTEXT (obj);
}

static AtkHypertext *
get_hypertext_from_path (const char *path, void *user_data)
{
  AtkObject *obj = spi_dbus_get_object (path);
  if (!obj)
    return NULL;
  return ATK_HYPERTEXT (obj);
}

static DBusMessage *
impl_getNLinks (DBusConnection * bus, DBusMessage * message, void *user_data)
{
  AtkHypertext *hypertext = get_hypertext (message);
  gint rv;
  DBusMessage *reply;

  if (!hypertext)
    return spi_dbus_general_error (message);
  rv = atk_hypertext_get_n_links (hypertext);
  reply = dbus_message_new_method_return (message);
  if (reply)
    {
      dbus_message_append_args (reply, DBUS_TYPE_INT32, &rv,
				DBUS_TYPE_INVALID);
    }
  return reply;
}

static DBusMessage *
impl_getLink (DBusConnection * bus, DBusMessage * message, void *user_data)
{
  AtkHypertext *hypertext = get_hypertext (message);
  DBusError error;
  dbus_int32_t linkIndex;
  AtkHyperlink *link;

  if (!hypertext)
    return spi_dbus_general_error (message);
  dbus_error_init (&error);
  if (!dbus_message_get_args
      (message, &error, DBUS_TYPE_INT32, &linkIndex, DBUS_TYPE_INVALID))
    {
      return SPI_DBUS_RETURN_ERROR (message, &error);
    }
  link = atk_hypertext_get_link (hypertext, linkIndex);
  return spi_dbus_return_object (message, ATK_OBJECT (link), FALSE);
}

static DBusMessage *
impl_getLinkIndex (DBusConnection * bus, DBusMessage * message,
		   void *user_data)
{
  AtkHypertext *hypertext = get_hypertext (message);
  DBusError error;
  dbus_int32_t characterIndex;
  dbus_int32_t rv;
  DBusMessage *reply;

  if (!hypertext)
    return spi_dbus_general_error (message);
  dbus_error_init (&error);
  if (!dbus_message_get_args
      (message, &error, DBUS_TYPE_INT32, &characterIndex, DBUS_TYPE_INVALID))
    {
      return SPI_DBUS_RETURN_ERROR (message, &error);
    }
  rv = atk_hypertext_get_link_index (hypertext, characterIndex);
  reply = dbus_message_new_method_return (message);
  if (reply)
    {
      dbus_message_append_args (reply, DBUS_TYPE_INT32, &rv,
				DBUS_TYPE_INVALID);
    }
  return reply;
}

static DRouteMethod methods[] = {
  {DROUTE_METHOD, impl_getNLinks, "getNLinks", "i,,o"},
  {DROUTE_METHOD, impl_getLink, "getLink", "i,linkIndex,i:o,,o"},
  {DROUTE_METHOD, impl_getLinkIndex, "getLinkIndex",
   "i,characterIndex,i:i,,o"},
  {0, NULL, NULL, NULL}
};

void
spi_initialize_hypertext (DRouteData * data)
{
  droute_add_interface (data, "org.freedesktop.accessibility.Hypertext",
			methods, NULL,
			(DRouteGetDatumFunction) get_hypertext_from_path,
			NULL);
};

File: pyatspi/hypertext.py
#Copyright (C) 2008 Codethink Ltd

#This library is free software; you can redistribute it and/or
#modify it under the terms of the GNU Lesser General Public
#License version 2 as published by the Free Software Foundation.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#You should have received a copy of the GNU Lesser General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

from interfaces import *
from accessible import Accessible
from factory import accessible_factory

__all__ = [
           "Hypertext",
          ]

#------------------------------------------------------------------------------

class Hypertext(Accessible):
        """
        An interface used for objects which implement linking between
        multiple resource or content locations, or multiple 'markers'
        within a single document. A Hypertext instance is associated
        with one or more Hyperlinks, which are associated with particular
        offsets within the Hypertext's included content.
        """

        def getLink(self, index):
                """
                Get one of the Hyperlinks associated with this Hypertext object,
                by index.
                @param : linkIndex
                an integer from 0 to getNLinks() - 1. 
                @return the Hyperlink in this Hypertext object.
                """
                func = self.get_dbus_method("getLink", dbus_interface=ATSPI_HYPERTEXT)
                return self._cache.create_accessible(self._app_name, func(index),
                                                     interfaces.ATSPI_HYPERTEXT)

        def getLinkIndex(self, character_index):
                """
                Get the hyperlink index, if any, associated with a particular
                character offset in the Hypertext object. For Hypertext implementors
                without textual content, all hyperlinks are associated with character
                offset '0'.
                @return the index of the Hyperlink associated with character
                offset characterIndex, or -1 if no Hyperlink is associated with
                that character offset.
                """
                func = self.get_dbus_method("getLinkIndex", dbus_interface=ATSPI_HYPERTEXT)
                return func(character_index)

        def getNLinks(self):
                """
                Query the hypertext object for the number of Hyperlinks it contains.
                @return the number of Hyperlinks associated with this Hypertext
                object, as a long integer.
                """
                func = self.get_dbus_method("getNLinks", dbus_interface=ATSPI_HYPERTEXT)
                return func()

# Register the accessible class with the factory.
accessible_factory.register_accessible_class(ATSPI_HYPERTEXT, Hypertext)

#END----------------------------------------------------------------------------
